pkg/services/ngalert/provisioning/config.go

```diff
@@ -48,3 +48,42 @@ func getLastConfiguration(ctx context.Context, orgID int64, store AMConfigStore)
 		version:          alertManagerConfig.ConfigurationVersion,
 	}, nil
 }
+
+type alertmanagerConfigStore interface {
+	Get(ctx context.Context, orgID int64) (*cfgRevision, error)
+	Save(ctx context.Context, revision *cfgRevision, orgID int64, afterSave func(ctx context.Context) error) error
+}
+
+type alertmanagerConfigStoreImpl struct {
+	store AMConfigStore
+	xact  TransactionManager
+}
+
+func (a alertmanagerConfigStoreImpl) Get(ctx context.Context, orgID int64) (*cfgRevision, error) {
+	return getLastConfiguration(ctx, orgID, a.store)
+}
+
+func (a alertmanagerConfigStoreImpl) Save(ctx context.Context, revision *cfgRevision, orgID int64, afterSave func(ctx context.Context) error) error {
+	serialized, err := serializeAlertmanagerConfig(*revision.cfg)
+	if err != nil {
+		return err
+	}
+	cmd := models.SaveAlertmanagerConfigurationCmd{
+		AlertmanagerConfiguration: string(serialized),
+		ConfigurationVersion:      revision.version,
+		FetchedConfigurationHash:  revision.concurrencyToken,
+		Default:                   false,
+		OrgID:                     orgID,
+	}
+	return a.xact.InTransaction(ctx, func(ctx context.Context) error {
+		err = PersistConfig(ctx, a.store, &cmd)
+		if err != nil {
+			return err
+		}
+		err = afterSave(ctx)
```

Comment:
Since there is only one database call here, this object doesn't technically need to create a transaction here.

And, since transactions are propagated through the context, can we just leave the transaction handling to the caller? That would let us drop the `afterSave` hook entirely, since the caller would own the transaction and can inject any extra code it likes, and only if it wants to do so, making a more clear abstraction.

Reply:
The idea is to completely hide the implementation details from the domain logic in service. So service just takes care of  the changes and manages stores (read and write). This will help us down the road when we use Cloud Alertmanager which likely won't use db transaction.